week-03/assessment/solar-farm/solar-farm/src/main/java/learn/solar/domain/PanelService.java

```diff
@@ -0,0 +1,84 @@
+import learn.solar.data.PanelRepository;
+import learn.solar.models.Material;
+import learn.solar.models.Panel;
+
+import java.util.List;
+
+// panel validation/rules
+public class PanelService {
+    private PanelRepository repository;
+
+    public PanelService(PanelRepository repository) {
+        this.repository = repository;
+    }
+
+    public List<Panel> findBySection(String section) throws DataException {
+        return repository.findBySection(section);
+    }
+
+    //validate, then add via repository
+    public PanelResult add(Panel panel) throws DataException {
+        PanelResult result = validate(panel);
+        if (!result.isSuccess()) {
+            return result;
+        }
+        panel = repository.add(panel);
+        result.setPanel(panel);
+        return result;
+    }
+
+    //validate, then update via repository
+    public PanelResult update(Panel panel) throws DataException {
+        PanelResult result = validate(panel);
+        if (!result.isSuccess()) {
+            return result;
+        }
+        boolean updated = repository.update(panel);
+        if (updated) {
+            result.setPanel(panel);
+        }
+        return result;
+    }
+        //pass thru to repository
+    public PanelResult deleteById(int id) throws DataException {
+        PanelResult result = new PanelResult();
+        if(!repository.deleteById(id)){
+            result.addMessage(String.format("Log entry with id: %s does not exist", id));
+        }
+        return result;
+    }
+
+    private PanelResult validate(Panel panel) {
+        PanelResult result = new PanelResult();
+
+        if (panel == null) {
+            result.addMessage("Panel entry cannot be null.");
+            return result;
+        }
+
+        if (panel.getSection() == null || panel.getSection().isBlank()) {
+            result.addMessage("Panel section is required.");
+            return result;
+        }////- **Row** is a positive number less than or equal to 250.
+        if (panel.getRow() < 0 || panel.getRow() > 250) {
+            result.addMessage("Panel Row is required.");
+            return result;
+        } //- **Column** is a positive number less than or equal to 250.
+        if (panel.getColumn() < 0 || panel.getColumn() > 250) {
+            result.addMessage("Description is required.");
+            return result;
+        } // **Year Installed** must be in the past.
+        if (panel.getInstallationYear() >= 2023) {
+            result.addMessage("Installation year must be before current year (2023).");
+            return result;
+        } //??? **Material** is required and can only be one of the five materials listed.
+        if (panel.getMaterial() == null) {
+            result.addMessage("Material is required and must be one of five materials listed: [multicrystalline silicon, monocrystalline silicon, amorphous silicon, cadmium telluride, or copper indium gallium selenide].");
+            return result;
+        }
```

Comment:
need to validate for duplicate panel here as well - cannot be the same section, row, column as another panel